events/controller.go

```diff
@@ -8,48 +8,50 @@ import (
 
 	pmetrics "github.com/awslabs/operatorpkg/metrics"
 	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/singleton"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/utils/clock"
 	controllerruntime "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
-	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 type Controller[T client.Object] struct {
-	gvk        schema.GroupVersionKind
-	startTime  time.Time
-	kubeClient client.Client
-	EventCount pmetrics.CounterMetric
+	gvk               schema.GroupVersionKind
+	startTime         time.Time
+	kubeClient        client.Client
+	EventCount        pmetrics.CounterMetric
+	EventWatchChannel <-chan watch.Event
 }
 
-func NewController[T client.Object](client client.Client, clock clock.Clock) *Controller[T] {
+func NewController[T client.Object](client client.Client, clock clock.Clock, channel <-chan watch.Event) *Controller[T] {
```

Comment:
Can you build this in a way that you don't require the caller to start the watch themselves and have to implement the channel?